src/server/qgsserverprojectutils.cpp

```diff
@@ -314,9 +315,103 @@ QStringList QgsServerProjectUtils::wmsOutputCrsList( const QgsProject &project )
   return crsList;
 }
 
-QString QgsServerProjectUtils::wmsServiceUrl( const QgsProject &project )
+QString QgsServerProjectUtils::header( const QgsServerRequest &request, const QString header )
 {
-  return project.readEntry( QStringLiteral( "WMSUrl" ), QStringLiteral( "/" ), "" );
+  QString result = request.header( header );
+  if ( result.isEmpty() ) {
+    result = qgetenv( QStringLiteral( "HTTP_%1" ).arg(
```

Comment:
These new settings should probably be used through the `QgsServerSettings` class. What do you think?

Reply:
This is a header not a settings then not this one but the ``SERVICE_URL`` should probably.

Reply:
Actually, I don't like having `qgetenv` (or similar) dispersed everywhere in QGIS Server source code. We're not there yet, but all environment variables should be centralized in a dedicated class (or subclasses for variables specialized for a given service like QGIS_SERVER_WMS_...).

Reply:
I refactor the pull request according your comment, then I move all the `qgetenv` in `QgsServerSettings` or in `QgsFcgiServerResponse` :-)

Reply:
@pblottiere up :-)